replication/json_binary_test.go

```diff
@@ -0,0 +1,403 @@
+	testCases := []struct {
+		name            string
+		value           float64
+		withTrailing    string // Expected output with FloatWithTrailingZero
+		withoutTrailing string // Expected output with regular float64
+	}{
+		{
+			name:            "whole number 5.0",
+			value:           5.0,
+			withTrailing:    "5.0",
+			withoutTrailing: "5",
+		},
+		{
+			name:            "zero",
+			value:           0.0,
+			withTrailing:    "0.0",
+			withoutTrailing: "0",
+		},
+		{
+			name:            "negative whole number",
+			value:           -42.0,
+			withTrailing:    "-42.0",
+			withoutTrailing: "-42",
+		},
+		{
+			name:            "decimal number (no difference)",
+			value:           3.14,
+			withTrailing:    "3.14",
+			withoutTrailing: "3.14",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Test FloatWithTrailingZero
+			trailingResult, err := json.Marshal(FloatWithTrailingZero(tc.value))
+			require.NoError(t, err)
+			require.Equal(t, tc.withTrailing, string(trailingResult))
+
+			// Test regular float64
+			regularResult, err := json.Marshal(tc.value)
+			require.NoError(t, err)
+			require.Equal(t, tc.withoutTrailing, string(regularResult))
+
+			// Verify they're different for whole numbers
+			if tc.value == float64(int(tc.value)) {
+				require.NotEqual(t, string(trailingResult), string(regularResult),
+					"FloatWithTrailingZero and regular float64 should produce different output for whole numbers")
+			}
+		})
+	}
+}
+
+func TestJsonBinaryDecoder_decodeDoubleWithTrailingZero(t *testing.T) {
+	// Test the decodeDoubleWithTrailingZero method directly
+	decoder := &jsonBinaryDecoder{
+		useFloatWithTrailingZero: true,
+	}
+
+	// Test data representing float64 in binary format (little endian)
+	// 5.0 as IEEE 754 double precision: 0x4014000000000000
+	testData := []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x14, 0x40}
+
+	result := decoder.decodeDoubleWithTrailingZero(testData)
+	require.NoError(t, decoder.err)
+
+	// Verify the result is FloatWithTrailingZero type
+	require.IsType(t, FloatWithTrailingZero(0), result)
+	require.Equal(t, FloatWithTrailingZero(5.0), result)
+
+	// Test JSON marshaling
+	jsonBytes, err := json.Marshal(result)
+	require.NoError(t, err)
+	require.Equal(t, "5.0", string(jsonBytes))
+}
+
+func TestJsonBinaryDecoder_decodeValue_JSONB_DOUBLE(t *testing.T) {
+	tests := []struct {
+		name                     string
+		useFloatWithTrailingZero bool
+		expectedType             interface{}
+		expectedJSONString       string
+	}{
+		{
+			name:                     "with trailing zero enabled",
+			useFloatWithTrailingZero: true,
+			expectedType:             FloatWithTrailingZero(0),
+			expectedJSONString:       "5.0",
+		},
+		{
+			name:                     "with trailing zero disabled",
+			useFloatWithTrailingZero: false,
+			expectedType:             float64(0),
+			expectedJSONString:       "5",
+		},
+	}
+
+	// 5.0 as IEEE 754 double precision in little endian
```

Comment:
nits: I think you can use https://pkg.go.dev/math#Float64bits and `binary.BigEndian.PutUint64`. This line with comment is also good enough.

Reply:
Updated to use the libraries instead of hardcoding the binary representation.